File: pachinko/submissions/accepted/pachinko_approximation.cpp
#include <bits/stdc++.h>
using namespace std;

const int M = 100;

/**
 * Return the expected length of a path that starts at node 1 in the pachinko.
 * 
 * N: number of nodes in the pachinko
 * S: starting node of the flipper
 * E: ending node of the flipper
 * P: list of parent nodes in the pachinko
 */
double solve(int N, int S, int E, vector<int> P) {
    // Let's make it 0-based
    --S, --E;
    for (int& p : P) --p;
    vector<vector<int>> g(N);
    for (int i = 0; i < N - 1; ++i) {
        g[P[i]].push_back(i + 1);
    }
    g[S].push_back(E);
    vector<int> visited(N, 0);

    auto expected = [&](int i, auto&& self) -> double {
        if (++visited[i] == M) return 0;
        double ans = 0;
        for (int j : g[i]) {
            ans += self(j, self);
        }
        ans /= double(g[i].size());
        return ans;
    };

    return expected(0, expected);
}

int main() {
    int T;
    cin >> T;
    while (T--) {
        int N, S, E;
        cin >> N >> S >> E;
        vector<int> P(N - 1);
        for (int i = 0; i < N - 1; ++i) {
            cin >> P[i];
        }
        cout << fixed << setprecision(6) << solve(N, S, E, P) << '\n';
    }
    return 0;
}

File: pachinko/submissions/accepted/pachinko_approximation.py
from collections import deque

M = 100


def solve(N: int, S: int, E: int, P: list[int]) -> int:
    """
    Return the expected length of a path that starts at node 1 in the pachinko.

    N: number of nodes in the pachinko
    S: starting node of the flipper
    E: ending node of the flipper
    P: list of parent nodes in the pachinko
    """
    P = [u - 1 for u in P]
    S -= 1
    E -= 1
    g = [list() for _ in range(N * M)]
    # Build M copies of the tree
    for j in range(M):
        for i in range(N - 1):
            g[N * j + P[i]].append(N * j + i + 1)
    # Build M - 1 copies of the flipper
    for j in range(M - 1):
        g[N * j + S].append(N * (j + 1) + E)

    # Toposort
    topo = []
    q = deque()
    indeg = [0 for i in range(len(g))]
    for u in range(len(g)):
        for v in g[u]:
            indeg[v] += 1
    for u in range(len(indeg)):
        if indeg[u] == 0:
            q.append(u)

    while q:
        u = q.popleft()
        topo.append(u)
        for v in g[u]:
            indeg[v] -= 1
            if indeg[v] == 0:
                q.append(v)

    answers = [0 for i in range(len(topo))]

    for u in topo[::-1]:
        for v in g[u]:
            answers[u] += 1 + answers[v]
        if g[u]:
            answers[u] /= len(g[u])

    return answers[0]


def main():
    T = int(input())
    for _ in range(T):
        N, S, E = [int(x) for x in input().split()]
        P = [int(x) for x in input().split()]
        print(solve(N, S, E, P))


if __name__ == '__main__':
    main()
